### PR TITLE
Web Hooks Agent

### PR DESCRIPTION
**Pull request by mejt**

### CODE REVIEW

**Comment by adrianpietka** [comment]
@zeulus Can you make a code review?

**Comment by mejt** [comment]
Please wait for new version.

**Comment by zeulus** [comment]
Okay, I'll wait

**Comment by mejt** [comment]
I changed this agent to more flexibility. Michał will add better routing and support for Bitbucket.

File: agents/webhooks-agent/public/index.php [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+require_once('../vendor/autoload.php');
+
+use FP\Larmo\Agents\WebHookAgent\Packet;
+use FP\Larmo\Agents\WebHookAgent\Request;
+use FP\Larmo\Agents\WebHookAgent\Services;
+
+header("Content-type: application/json; charset=utf-8");
+
+if(Request::isPostMethod()) {
```
Comment (adrianpietka) [inline agents/webhooks-agent/public/index.php]: Look on my PoC solution for creating a different services: https://gist.github.com/adrianpietka/e1c361e82c4d09a8967c

File: agents/webhooks-agent/src/Metadata.php [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+namespace FP\Larmo\Agents\WebHookAgent;
+
+class Metadata {
+    private $timestamp;
+    private $source;
+    private $authinfo;
+
+    public function __construct($timestamp = null) {
+        $this->timestamp = $timestamp ? $timestamp : time();
+        $this->source = "github";
```
Comment (adrianpietka) [inline agents/webhooks-agent/src/Metadata.php]: You can't hardcode name of source in metadata

File: agents/webhooks-agent/src/Request.php [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    public static function getUri() {
+        return $_SERVER["REQUEST_URI"];
+    }
+
+    public static function getPostData() {
+        return file_get_contents('php://input');
+    }
+
+    public static function getValueFromHeaderByKey($key) {
+        foreach($_SERVER as $headerKey => $headerValue) {
```
Comment (adrianpietka) [inline agents/webhooks-agent/src/Request.php]:
```php
return array_key_exists($key, $_SERVER) ? $_SERVER[$key] : true;
```

??

File: agents/webhooks-agent/src/Services/Github/GithubData.php [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    private $data;
+
+    public function __construct($data) {
+        $this->data = $this->prepareData($data);
+    }
+
+    private function prepareData($data) {
+        $eventType = strtolower(Request::getValueFromHeaderByKey("HTTP_X_GITHUB_EVENT"));
+
+        if($eventType) {
```
Comment (adrianpietka) [inline agents/webhooks-agent/src/Services/Github/GithubData.php]: Try replace "switch" to factory (similar to PoC code in first comment)

**Comment by mejt** [comment]
I pushed commit with changes my changes and @mikron-ia. I added some unit tests for Github events and I changed code style.

Additionally @mikron-ia did a lot of things: 
- rewrote routing
- added exceptions
- basic support for bitbucket

File: agents/webhooks-agent/.gitignore [hunk 5]
```diff
@@ -0,0 +1,2 @@
+.idea/
```
Comment (zeulus) [inline agents/webhooks-agent/.gitignore]: Oups... add it to gitignore next time :)

File: agents/webhooks-agent/src/Request.php [hunk 6]
```diff
@@ -0,0 +1,36 @@
+<?php
+
+namespace FP\Larmo\Agents\WebHookAgent;
+
+class Request
+{
+    public static function isPostMethod()
+    {
+        if ($_SERVER['REQUEST_METHOD'] === 'POST') {
```
Comment (adrianpietka) [inline agents/webhooks-agent/src/Request.php]: Change it to:

```php
return $_SERVER['REQUEST_METHOD'] === 'POST';
```

File: agents/webhooks-agent/public/index.php [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+use FP\Larmo\Agents\WebHookAgent\Packet;
+use FP\Larmo\Agents\WebHookAgent\Request;
+use FP\Larmo\Agents\WebHookAgent\Routing;
+use FP\Larmo\Agents\WebHookAgent\Services;
+use FP\Larmo\Agents\WebHookAgent\Services\ServiceFactory;
+use FP\Larmo\Agents\WebHookAgent\Exceptions\MethodNotAllowedHttpException;
+use FP\Larmo\Agents\WebHookAgent\Exceptions\ServiceNotFoundException;
+use FP\Larmo\Agents\WebHookAgent\Exceptions\EventTypeNotFoundException;
+
+header("Content-type: application/json; charset=utf-8");
```
Comment (adrianpietka) [inline agents/webhooks-agent/public/index.php]: @mejt please choose " or '. I suggest single quote :rabbit:

File: agents/webhooks-agent/public/index.php [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+        throw new MethodNotAllowedHttpException;
+    }
+
+    /* Retrieve data from HTTP request */
+    $uri = Request::getUri();
+    $postData = json_decode(Request::getPostData());
+
+    /* Create appropriate service */
+    $routing = new Routing($uri);
+    $service = ServiceFactory::create($routing->getSourceIdentifier(),$postData);
```
Comment (dalailomo) [inline agents/webhooks-agent/public/index.php]: "there MUST be one space after each comma"

http://www.php-fig.org/psr/psr-2/#4.4.-method-arguments

File: agents/webhooks-agent/src/Packet.php [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+class Packet
+{
+    private $messages;
+    private $metadata;
+    private $packetArray;
+
+    public function __construct(ServiceDataInterface $service)
+    {
+        $this->metadata = new Metadata($service->getServiceName());
```
Comment (adrianpietka) [inline agents/webhooks-agent/src/Packet.php]: I propose use a dependency injection for Metadata object

File: agents/webhooks-agent/src/Exceptions/ServiceNotFoundException.php [hunk 10]
```diff
@@ -0,0 +1,9 @@
+<?php
+
+namespace FP\Larmo\Agents\WebHookAgent\Exceptions;
+
+
```
Comment (dalailomo) [inline agents/webhooks-agent/src/Exceptions/ServiceNotFoundException.php]: we should be consistent with this white spaces... or 1, or 2, but not mixed :)

File: agents/webhooks-agent/src/Packet.php [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+        $this->messages = $service->getData();
+        $this->packetArray = $this->preparePacket();
+    }
+
+    private function preparePacket()
```
Comment (dalailomo) [inline agents/webhooks-agent/src/Packet.php]: for me, what this method does sounds looks to me more like "getting" something than "preparing" something, IMO :)

And also maybe this array can be created in the constructor, and we do not need to create private properties `$this->messages` and `$this->metadata` (these ones are only used in this private method...)

File: agents/webhooks-agent/src/Packet.php [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+        $packetJson = json_encode($this->packetArray);
+        $this->saveDataToFile($packetJson);
+        $this->saveDataAsLog($packetJson);
+
+        return $packetJson;
+    }
+
+    private function saveDataToFile($packet)
+    {
+        $myFile = fopen("output.json", "w+");
```
Comment (dalailomo) [inline agents/webhooks-agent/src/Packet.php]: hardcoded value.

Maybe we should consider using http://php.net/manual/en/class.splfileobject.php ;)

File: agents/webhooks-agent/src/Request.php [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    public static function getPostData()
+    {
+        return file_get_contents('php://input');
+    }
+
+    public static function getValueFromHeaderByKey($key)
+    {
+        foreach ($_SERVER as $headerKey => $headerValue) {
```
Comment (dalailomo) [inline agents/webhooks-agent/src/Request.php]: What about doing just something like...
```
return (isset($_SERVER[$key])) ? $_SERVER[$key] : null;
```

File: agents/webhooks-agent/src/Request.php [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+            if ($key === $headerKey) {
+                return $headerValue;
+            }
+        }
+
+        return true;
```
Comment (dalailomo) [inline agents/webhooks-agent/src/Request.php]: `true`? Oo.

File: agents/webhooks-agent/src/Routing.php [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+    private $uri;
+
+    public function __construct($uri)
+    {
+        $this->uri = $uri;
+    }
+
+    public function getSourceIdentifier()
+    {
+        $cleansedUri = trim(str_replace("http://", "", $this->uri));
```
Comment (dalailomo) [inline agents/webhooks-agent/src/Routing.php]: should we take into account `https://`?

Comment (dalailomo) [inline agents/webhooks-agent/src/Routing.php]: And also to extract the last segment of a URI, you can try using `basename()`

http://php.net/manual/en/function.basename.php

File: agents/webhooks-agent/src/Request.php [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+
+        return false;
+    }
+
+    public static function getUri()
+    {
+        return $_SERVER["REQUEST_URI"];
+    }
+
+    public static function getPostData()
```
Comment (adrianpietka) [inline agents/webhooks-agent/src/Request.php]: It's not a "real" POST data, rename it to "getPayload".

File: agents/webhooks-agent/src/Request.php (hunk 13 above)
Comment (adrianpietka) [inline agents/webhooks-agent/src/Request.php]: @dalailomo :+1: in previous CR I suggested:

```php
return array_key_exists($key, $_SERVER) ? $_SERVER[$key] : true;
```

Why array_key_exists() not isset()? Isset return "false" when value is null.

File: agents/webhooks-agent/src/Routing.php (hunk 15 above)
Comment (adrianpietka) [inline agents/webhooks-agent/src/Routing.php]: or use: http://php.net/manual/en/function.parse-url.php

File: agents/webhooks-agent/src/Services/Bitbucket/BitbucketData.php [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+        $this->serviceName = 'bitbucket';
+        $this->data = $this->prepareData($data);
+    }
+
+    private function prepareData($data)
+    {
+        $eventType = strtolower(Request::getValueFromHeaderByKey("HTTP_X_EVENT_KEY"));
+
+        if (!empty($eventType)) {
+            switch ($eventType) {
```
Comment (adrianpietka) [inline agents/webhooks-agent/src/Services/Bitbucket/BitbucketData.php]: Remember, each switch statement sounds like Strategy Pattern ;-)

File: agents/webhooks-agent/src/Services/Bitbucket/BitbucketData.php [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+    {
+        $this->serviceName = 'bitbucket';
+        $this->data = $this->prepareData($data);
+    }
+
+    private function prepareData($data)
+    {
+        $eventType = strtolower(Request::getValueFromHeaderByKey("HTTP_X_EVENT_KEY"));
+
+        if (!empty($eventType)) {
```
Comment (dalailomo) [inline agents/webhooks-agent/src/Services/Bitbucket/BitbucketData.php]: I would do...
```
if (empty($eventType)) {
  throw new \InvalidArgumentException;
}

// rest of the code
```

imo looks more readable :)

File: agents/webhooks-agent/src/Services/Bitbucket/BitbucketData.php [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+
+    public function __construct($data)
+    {
+        $this->serviceName = 'bitbucket';
+        $this->data = $this->prepareData($data);
+    }
+
+    private function prepareData($data)
+    {
+        $eventType = strtolower(Request::getValueFromHeaderByKey("HTTP_X_EVENT_KEY"));
```
Comment (adrianpietka) [inline agents/webhooks-agent/src/Services/Bitbucket/BitbucketData.php]: and now you BitbucketData class is depends on you implementation of Request class. It's not the best way, every time when object need instance of other object then we should use a dependency injection.

additional: look on "low copuling" in GRASP principles: https://en.wikipedia.org/wiki/GRASP_(object-oriented_design)#Low_Coupling

File: agents/webhooks-agent/src/Request.php (hunk 13 above)
Comment (dalailomo) [inline agents/webhooks-agent/src/Request.php]: yep, you re right :+1:

File: agents/webhooks-agent/src/Services/Github/GithubData.php [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+        $this->serviceName = 'github';
+        $this->data = $this->prepareData($data);
+    }
+
+    private function prepareData($data)
+    {
+        $eventType = strtolower(Request::getValueFromHeaderByKey("HTTP_X_GITHUB_EVENT"));
+
+        if ($eventType) {
+            switch ($eventType) {
```
Comment (adrianpietka) [inline agents/webhooks-agent/src/Services/Github/GithubData.php]: Look up - switch -> Strategy Pattern.

File: agents/webhooks-agent/src/Services/ServiceFactory.php [hunk 21]
```diff
@@ -0,0 +1,36 @@
+<?php
+/**
```
Comment (adrianpietka) [inline agents/webhooks-agent/src/Services/ServiceFactory.php]: This comment is unnecessary

File: agents/webhooks-agent/src/Services/ServiceFactory.php [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+
+class ServiceFactory
+{
+    static public function create($serviceName, $data)
+    {
+        try {
+            /* Caution: full namespace path is necessary for class_exists() to work correctly */
+            $serviceClass = "\\FP\\Larmo\\Agents\\WebHookAgent\\Services\\" . ucfirst($serviceName) . "\\" . ucfirst($serviceName) . "Data";
+
+            if (!class_exists($serviceClass)) {
```
Comment (adrianpietka) [inline agents/webhooks-agent/src/Services/ServiceFactory.php]: else block is unnecessary. More readable are:

```php
if (class_exists($serviceClass)) {
 return new $serviceClass($data);
}

throw new ServiceNotFoundException;
```

or:

```php
if (!class_exists($serviceClass)) {
 throw new ServiceNotFoundException;
}

return new $serviceClass($data);
```

File: agents/webhooks-agent/src/Services/Bitbucket/Events/Push.php [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+
+use FP\Larmo\Agents\WebHookAgent\Services\Bitbucket\EventInterface;
+
+/**
+ * Class Push
+ * @package FP\Larmo\Agents\WebHookAgent\Services\Bitbucket\Events
+ *
+ * Handles pushed commits
+ * Caution: BitBucket is not providing entirety of data in push report, only the last commit and information regarding previous HEAD
+ * Therefore, this event will not report all commits in a push, just the most recent one; this appears to be intentional behaviour of BitBucket API
```
Comment (dalailomo) [inline agents/webhooks-agent/src/Services/Bitbucket/Events/Push.php]: http://www.php-fig.org/psr/psr-2/#2.3.-lines

I don't know how long it is, its just to remind to not put more than 120 :)

File: agents/webhooks-agent/src/Services/Bitbucket/Events/Push.php [hunk 24]
```diff
@@ -0,0 +1,60 @@
...
+ */
+class Push implements EventInterface
```
Comment (dalailomo) [inline agents/webhooks-agent/src/Services/Bitbucket/Events/Push.php]: mmmm I think we should not use verbs to name clases... imo

File: agents/webhooks-agent/src/Services/Github/EventInterface.php [hunk 25]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+namespace FP\Larmo\Agents\WebHookAgent\Services\Github;
+
+interface EventInterface
+{
+    /**
+     * @param $data
```
Comment (dalailomo) [inline agents/webhooks-agent/src/Services/Github/EventInterface.php]: data type should be indicated :)

File: agents/webhooks-agent/src/Services/ServiceFactory.php [hunk 26]
```diff
@@ -0,0 +1,36 @@
...
+                throw new ServiceNotFoundException;
+            } else {
+                return new $serviceClass($data);
+            }
+        } catch (EventTypeNotFoundException $e) {
+            throw new EventTypeNotFoundException;
```
Comment (adrianpietka) [inline agents/webhooks-agent/src/Services/ServiceFactory.php]: Should rethrow the same exception:

```php
} catch (EventTypeNotFoundException $e) {
   throw $e;
}
```

File: agents/webhooks-agent/src/Services/ServiceFactory.php [hunk 27]
```diff
@@ -0,0 +1,36 @@
...
+        } catch (ServiceNotFoundException $e) {
+            throw new ServiceNotFoundException;
```
Comment (adrianpietka) [inline agents/webhooks-agent/src/Services/ServiceFactory.php]: The same to previous comment.

File: agents/webhooks-agent/src/Services/Bitbucket/Events/Push.php (hunk 24 above)
Comment (mikron-ia) [inline agents/webhooks-agent/src/Services/Bitbucket/Events/Push.php]: I believe this is 'push' as a noun, as 'this is the push containing the commits'. I would leave it as it is, as this is how git-hosters call this event.

**Comment by mejt** [comment]
I pushed composer.lock file because it is required for deploying to Heroku.

**Pull request merged**